### PR TITLE
First draft: Connect, set-value! and get-value (as a string)

### PR DESCRIPTION
**Pull request by viebel**

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-  :profiles {:dev {:dependencies [
-                [midje]
-              ]}
-                                  }
-  :parent-project {:coords [com.cycognito/cyco-parent-clj "0.0.31"]
+                 [com.cycognito/sea-cucumber]
+                 [clj-time]
+                 [com.stuartsierra/component]
+                 [cbass :exclusions [[org.apache.hbase/hbase-shaded-client]]]
+                 [com.google.cloud.bigtable/bigtable-hbase-1.x :exclusions [[org.slf4j/slf4j-log4j12]]]]
```
Comment (itaysabato) [inline project.clj:9]: Please make sure that the dependencies (and everything else) is based on the "best" variation out of all the example projects. Dennis was the first, so I tend to suspect it is the least up to date - but this is not necessarily so.

Please make sure to compare the variations and verify with the author when you are not sure.

File: src/hamurai/component.clj [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    (not connection) (assoc :connection (connect! config))))
+
+(defrecord BigTable [config]
+  component/Lifecycle
+
+  (start [big-table]
+    (start* big-table))
+
+  (stop [big-table]
+    (stop* big-table)))
```
Comment (itaysabato) [inline src/hamurai/component.clj:34]: I find this a bit artificial. If you really want to make the component of Hamurai decoupled from the implementation take it all the way and create a namespace that does only that - wrap the implementation with a component.

File: src/hamurai/component.clj [hunk 3]
```diff
@@ -0,0 +1,40 @@
+(ns hamurai.component
+  (:require
+   [cbass :refer [pack-un-pack]]
+   [com.stuartsierra.component :as component])
+  (:import (com.google.cloud.bigtable.hbase BigtableConfiguration)
+           (org.apache.hadoop.hbase.client Connection)))
+
+(defn connect! [{:keys [big-table-project big-table-instance]} ]
+  (pack-un-pack {:p identity :u identity}) ;; TODO:  warning - global state change for this library
```
Comment (itaysabato) [inline src/hamurai/component.clj:9]: Please add a more elaborate comment about what this does and why we do it here.

Comment (itaysabato) [reply src/hamurai/component.clj:9]: If this is indeed what we need to do, I am willing to bet there is a better way to globally monkey-patch a 3rd party namespace.

File: src/hamurai/component.clj [hunk 4]
```diff
@@ -0,0 +1,40 @@
+(ns hamurai.component
+  (:require
+   [cbass :refer [pack-un-pack]]
+   [com.stuartsierra.component :as component])
+  (:import (com.google.cloud.bigtable.hbase BigtableConfiguration)
+           (org.apache.hadoop.hbase.client Connection)))
+
+(defn connect! [{:keys [big-table-project big-table-instance]} ]
```
Comment (itaysabato) [inline src/hamurai/component.clj:8]: Perhaps better:
```clj
; id or name?
[{{:keys [project-id instance-id]} :bigtable}]
```

File: src/hamurai/component.clj [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+(defn stop* [{:keys [connection] :as big-table}]
+  (when connection
+    (disconnect! connection))
+  (dissoc big-table :connection))
+
+(defn start* [{:keys [connection config] :as big-table}]
+  (cond-> big-table
+    (not connection) (assoc :connection (connect! config))))
+
+(defrecord BigTable [config]
```
Comment (itaysabato) [inline src/hamurai/component.clj:27]: In official docs and marketing it is consistently called `Bigtable` - let's adopt this.

File: src/hamurai/component.clj [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  component/Lifecycle
+
+  (start [big-table]
```
Comment (itaysabato) [inline src/hamurai/component.clj:30]: Due to the above - `bigtable`
We can also choose to default to `bt`.

File: src/hamurai/component.clj [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    (start* big-table))
+
+  (stop [big-table]
+    (stop* big-table)))
+
+
+
+(comment
+  (def connection (connect! {:big-table-project "cyco-main"
+                             :big-table-instance "qa-instance"})))
```
Comment (itaysabato) [inline src/hamurai/component.clj:40]: Let's start getting use to putting these types of comments in the test files - wouldn't you agree it's better?
It also make you create a test, even if it only has comments to begin with :)

File: src/hamurai/core.clj [hunk 8]
```diff
@@ -0,0 +1,48 @@
+(ns hamurai.core
+  (:require [hamurai.utils :refer [serialize ensure-keyword ]]
```
Comment (itaysabato) [inline src/hamurai/core.clj:2]: Please start using an auto-formater :)

File: src/hamurai/core.clj [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            [cbass :refer [find-by store]]
+            [sea-cucumber.core :refer [doapm]]
+            [clj-time.core :refer [now]]))
+
+(defn set-value! [conn {:keys [table-name row-key family column value]
+                  :or {family "default"}}]
+  (doapm "Store a value on BigTable"
```
Comment (itaysabato) [inline src/hamurai/core.clj:9]: Please use the (normalized) function name for this.
I tried to implement a macro that does this automatically, e.g.
```clj
(defn-inst get-value [conn {:keys [table-name row-key family column]
                       :or {family "default"}}]
  (-> (find-by conn table-name row-key family #{column})
             (get (ensure-keyword column))
             String.))
```

But it turns out `defn` does quite a lot more than just `(def name (fn ...))`

File: src/hamurai/core.clj [hunk 10]
```diff
@@ -0,0 +1,48 @@
+(ns hamurai.core
+  (:require [hamurai.utils :refer [serialize ensure-keyword ]]
+            [cbass :refer [find-by store]]
+            [sea-cucumber.core :refer [doapm]]
+            [clj-time.core :refer [now]]))
+
+(defn set-value! [conn {:keys [table-name row-key family column value]
```
Comment (itaysabato) [inline src/hamurai/core.clj:7]: I prefer:
```clj
(defn set-column-value! 
   ([bt {:keys [table]} row column value])
   ([bt {:keys [table]} row column-family column value]))
```

And later add something like:
```clj
(defn set-row-values!
   ([bt {:keys [table]} row family->column->value]))
```

File: src/hamurai/core.clj [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                  :or {family "default"}}]
+  (doapm "Store a value on BigTable"
+         (store conn table-name row-key family {column (serialize value)})))
+
+(defn get-value [conn {:keys [table-name row-key family column]
```
Comment (itaysabato) [inline src/hamurai/core.clj:12]: `get-column-value` with similar signature to above suggestion.

File: src/hamurai/core.clj (hunk 9 above)
Comment (viebel) [reply src/hamurai/core.clj:9]: I wrote an article about how to write a `defn` like macro https://blog.klipse.tech/clojure/2019/03/08/spec-custom-defn.html

File: src/hamurai/utils.clj [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  (.getBytes (unparse iso8601-formatter t)))
+
+(defn time? [x]
+  (instance? DateTime x))
+
+(defn serialize [x]
+  (cond
+    (time? x) (time-serialize x)
+    (string? x) (.getBytes x)
+    :else (throw (ex-info "Cannot serialize value" {:value x
```
Comment (itaysabato) [inline src/hamurai/utils.clj:18]: Please support `bytes?`

File: src/hamurai/utils.clj [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+                                                    :type (type x)}))))
+
+(defn ensure-keyword [x]
```
Comment (itaysabato) [inline src/hamurai/utils.clj:21]: Gearhead?

File: src/hamurai/utils.clj (hunk 12 above)
Comment (viebel) [reply src/hamurai/utils.clj:18]: Of course. How could I have missed that?

File: project.clj (hunk 1 above)
Comment (viebel) [reply project.clj:9]: dennis, lester and combo do all the same

File: src/hamurai/component.clj (hunk 2 above)
Comment (viebel) [reply src/hamurai/component.clj:34]: I prefer not to have code embedded in a record. Even when it is simple. Tools like clj-kondo and scope-capute do not deal well with methods.
That's why I use this separation between `start` and `start*`

File: src/hamurai/component.clj (hunk 3 above)
Comment (viebel) [reply src/hamurai/component.clj:9]: Indeed there is a better way. https://github.com/tolitius/cbass#when-connecting

File: src/hamurai/component.clj (hunk 5 above)
Comment (viebel) [reply src/hamurai/component.clj:27]: Totally agree. Respecting the case is important. Did you know that one should write JavaScript and ClojureScript with a capital S?

File: src/hamurai/component.clj (hunk 6 above)
Comment (viebel) [reply src/hamurai/component.clj:30]: agree

File: src/hamurai/core.clj (hunk 8 above)
Comment (viebel) [reply src/hamurai/core.clj:2]: I have an auto formatter but it doesn't run automatically :)

File: src/hamurai/component.clj (hunk 7 above)
Comment (viebel) [reply src/hamurai/component.clj:40]: I prefer to have them in the source files. Even Clojure core teams use comment blocks in the sources. See https://github.com/clojure/clojure/blob/master/src/clj/clojure/inspector.clj for instance

File: src/hamurai/core.clj (hunk 10 above)
Comment (viebel) [reply src/hamurai/core.clj:7]: ok

File: src/hamurai/utils.clj (hunk 13 above)
Comment (viebel) [reply src/hamurai/utils.clj:21]: In fact, when starting to move `ensure-keyword` to gearhead, I discovered that `clojure.core/keyword` works also with keywords. So `ensure-keyword` is superfluous.

File: src/hamurai/core.clj (hunk 9 above)
Comment (viebel) [reply src/hamurai/core.clj]: Do we really need to normalize tag names?
In the DD docs, they say they replace all special characters with `_`.
https://docs.datadoghq.com/getting_started/tagging/#defining-tags

Comment (itaysabato) [reply src/hamurai/core.clj]: This talk about the key, not the value.

**Pull request merged**